File: easy/plus-one.py
# https://leetcode.com/problems/plus-one/

"""
Strategy:

Loop from right to left,
	- If you run into a value that is < 9: Increment it and return early
	- If you run into a 9, convert it into 0
At the end of the loop if the conditions above are not met, then it means the values were all 9's
Prepend a 1 at the beginning and return the array
"""
from typing import List

class Solution:
	def plusOne(self, digits: List[int]) -> List[int]:
		left = len(digits) - 1

		while left >= 0:
			if digits[left] < 9:
				digits[left] += 1
				return digits
			digits[left] = 0 
			left -= 1

		return [1] + digits

solution = Solution()
sample = [1, 4, 9]
res = solution.plusOne(sample)

print(res)
assert res == [1, 5, 0], "[1, 5, 0] Expected"
	

File: easy/plus-one.ts
// https://leetcode.com/problems/plus-one/ 
function plusOne(digits: number[]): number[] {
    // Start a loop from right to left looking for the first non-9 value you run to
    var idx = digits.length - 1;

    while(idx >= 0){
        // If you find the first non-9 value, increment it by 1 and return the array
        // Your business ends here
        if(digits[idx] < 9){
            digits[idx] += 1;
            return digits;
        }

        // Otherwise, it means you ran into a 9, convert it into zero and resume your loop
        digits[idx] = 0
        idx -= 1;
    }

    // If you reach the end of the loop and you havent returned anything
    // then it means you ran into a 999, 9999 situation, 
    // which were then converted to 0s, i.e 000 and 0000 respectively, 
    // therefore all you have to do here is prepend 1 at the beginning of the array
    // and return the whole array post mutation

    // I'm sure there's a better way to do this in TS
    var res = [1];

    for(var i = 0; i < digits.length; i+=1){
        res.push(0);
    }

    return res;
};
